pkg/certificate/mrc_reconciler.go

```diff
@@ -0,0 +1,184 @@
+package certificate
+
+import (
+	time "time"
+
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
+	"github.com/openservicemesh/osm/pkg/constants"
+)
+
+// UseCase is how the certificate is used. Each use case will have its own rotation mechanism, that is necessary to
+// determine if we can propagate the rotation forward.
+type UseCase string
+
+const (
+	// UseCaseValidatingWebhook maps to the certificate use case for the validating webhook.
+	UseCaseValidatingWebhook UseCase = "validatingWebhook"
+
+	// UseCaseMutatingWebhook maps to the certificate use case for the Mutating Webhook
+	UseCaseMutatingWebhook UseCase = "mutatingWebhook"
+
+	// UseCaseXDSControlPlane maps to the certificate use case for the XDS Control Plane
+	UseCaseXDSControlPlane UseCase = "xdsControlPlane"
+
+	// UseCaseSidecar maps to the certificate use case for the  sidecar service certs
+	UseCaseSidecar UseCase = "sidecar"
+
+	// UseCaseBootstrap maps to the certificate use case for the  Bootstrap secrets
+	UseCaseBootstrap UseCase = "bootstrap"
+
+	// UseCaseGateway maps to the certificate use case for the  Gateway ingress cert.
+	UseCaseGateway UseCase = "gateway"
+)
+
+func (uc UseCase) String() string {
+	return string(uc)
+}
+func (m *Manager) handleMRCEvent(event MRCEvent) error {
+	mrc := event.MRC
+	// TODO(5046): improve logging in this method.
+	if m.shouldUpdateMRCComponentStatus(mrc) {
+		if err := m.updateMRCComponentStatus(mrc); err != nil {
+			return err
+		}
+	} else if shouldUpdateMRCState(mrc) {
+		if err := m.updateMRCState(mrc); err != nil {
+			return err
+		}
+	} else if m.shouldSetIssuers(mrc) {
+		log.Info().Msgf("setting new certificate issuers on the Certificate Manager for MRC %s in stage %s", mrc.GetName(), mrc.Status.State)
+		return m.setIssuers(mrc)
+	}
+	return nil
+}
+
+// Component statuses determine if we're ready to transition the mrc to the next state.
+// TODO(5046): add unit tests for this, and other methods in here.
+func (m *Manager) shouldUpdateMRCComponentStatus(mrc *v1alpha2.MeshRootCertificate) bool {
+	if len(mrc.Status.Conditions) == 0 {
+		return false
+	}
+
+	// Has enough time passed to consider the component status as ready?
+	lastUpdate := mrc.Status.Conditions[len(mrc.Status.Conditions)-1].LastTransitionTime.Time
```

Comment:
Could you explain the condition indexing here? I think we might want to be checking the `queuedTransition` time (from Keith's proposal) for each `componentStatus`.